### PR TITLE
Adding capability to create ray cluster with serve support

### PR DESCRIPTION
**Pull request by blublinsky**
## Why are these changes needed?

Currently, RayCluster CRD does not provide capabilities to create a cluster that can be used for Ray Serve. The only way to create such a cluster is to use RayServe CRD. This PR overcomes this by creating a cluster supporting Ray serve if an annotation `"ray.io/enableAgentService": "true"` is added to the cluster. 

## Related issue number

closes https://github.com/ray-project/kuberay

## Checks

- [ x] I've made sure the tests are passing. 
- Testing Strategy
   - [x ] Unit tests
   - [ x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by blublinsky** [comment]
@kevin85421  I was waiting for a while to implement this and it turned out to be extremely simple. The amount of additional code is minimal, so I assume this is not a big deal.
@z103cb, please take a look as well. I added a apiserver/test/cluster/cluster/cluster_with_serve to be able to test it from the API server

File: ray-operator/controllers/ray/common/pod.go [hunk 1]
```diff
@@ -258,7 +258,10 @@ func initHealthProbe(probe *v1.Probe, rayNodeType rayv1.RayNodeType) {
 }
 
 // BuildPod a pod config
-func BuildPod(podTemplateSpec v1.PodTemplateSpec, rayNodeType rayv1.RayNodeType, rayStartParams map[string]string, headPort string, enableRayAutoscaler *bool, creator string, fqdnRayIP string) (aPod v1.Pod) {
+func BuildPod(podTemplateSpec v1.PodTemplateSpec, rayNodeType rayv1.RayNodeType, rayStartParams map[string]string, headPort string, enableRayAutoscaler *bool, creator string, fqdnRayIP string, enableAgentService bool) (aPod v1.Pod) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod.go:261]: The name of `enableAgentService` is confusing. The K8s service for dashboard agent has already been removed from KubeRay by PR #1207. I guess you might mean `enableServeService`?

File: ray-operator/controllers/ray/common/service.go [hunk 2]
```diff
@@ -151,20 +151,40 @@ func BuildHeadServiceForRayService(rayService rayv1.RayService, rayCluster rayv1
 	return service, nil
 }
 
-// BuildServeServiceForRayService builds the service for head node and worker nodes who have healthy http proxy to serve traffics.
+// BuildServeServiceForRayService builds the serve service for Ray service.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service.go:154]:
```suggestion
// BuildServeServiceForRayService builds the serve service for RayService.
```

File: ray-operator/controllers/ray/common/service.go [hunk 3]
```diff
@@ -151,20 +151,40 @@ func BuildHeadServiceForRayService(rayService rayv1.RayService, rayCluster rayv1
...
+	namespace := rayCluster.Namespace
+	if service {
+		name = rayService.Name
+		namespace = rayService.Namespace
+	}
+
 	labels := map[string]string{
-		RayServiceLabelKey:               rayService.Name,
-		RayClusterServingServiceLabelKey: utils.GenerateServeServiceLabel(rayService.Name),
+		RayServiceLabelKey:               name,
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service.go:174]: Is this label still needed if we create a service for RayCluster?

File: ray-operator/controllers/ray/common/service.go [hunk 4]
```diff
@@ -151,20 +151,40 @@ func BuildHeadServiceForRayService(rayService rayv1.RayService, rayCluster rayv1
...
+	return BuildServeService(rayService, rayCluster, true)
+}
+
+// BuildServeServiceForRayCluster builds the serve service for Ray cluster.
+func BuildServeServiceForRayCluster(rayCluster rayv1.RayCluster) (*corev1.Service, error) {
+	return BuildServeService(rayv1.RayService{}, rayCluster, false)
+}
+
+// BuildServeService builds the service for head node and worker nodes who have healthy http proxy to serve traffics.
+func BuildServeService(rayService rayv1.RayService, rayCluster rayv1.RayCluster, service bool) (*corev1.Service, error) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service.go:165]:
```suggestion
func BuildServeService(rayService rayv1.RayService, rayCluster rayv1.RayCluster, isRayService bool) (*corev1.Service, error) {
```

File: ray-operator/controllers/ray/common/service_test.go [hunk 5]
```diff
@@ -425,6 +448,32 @@ func TestBuildServeServiceForRayService(t *testing.T) {
...
+	svc, err := BuildServeServiceForRayCluster(*instanceForServeSvc)
+	assert.Nil(t, err)
+
+	actualResult := svc.Spec.Selector[RayClusterLabelKey]
+	expectedResult := string(instanceForServeSvc.Name)
+	if !reflect.DeepEqual(expectedResult, actualResult) {
+		t.Fatalf("Expected `%v` but got `%v`", expectedResult, actualResult)
+	}
+
+	actualLabel := svc.Labels[RayServiceLabelKey]
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service_test.go:461]: Maybe the K8s serve service for the RayCluster should not have this label.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 6]
```diff
@@ -337,6 +337,15 @@ func (r *RayClusterReconciler) rayClusterReconcile(ctx context.Context, request
 		}
 		return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
 	}
+	// only reconcile serve service is "ray.io/enableAgentService" annotation is set to true
+	if enableAgentServiceValue, exist := instance.Annotations[common.EnableAgentServiceKey]; exist && enableAgentServiceValue == common.EnableAgentServiceTrue {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:341]: `EnableAgentServiceKey` -> `EnableServeServiceKey`

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 7]
```diff
@@ -1022,7 +1064,12 @@ func (r *RayClusterReconciler) buildWorkerPod(instance rayv1.RayCluster, worker
 	autoscalingEnabled := instance.Spec.EnableInTreeAutoscaling
 	podTemplateSpec := common.DefaultWorkerPodTemplate(instance, worker, podName, fqdnRayIP, headPort)
 	creatorName := getCreator(instance)
-	pod := common.BuildPod(podTemplateSpec, rayv1.WorkerNode, worker.RayStartParams, headPort, autoscalingEnabled, creatorName, fqdnRayIP)
+	// Check whether serve is enabled and add serve label
+	serveLabel := false
+	if enableAgentServiceValue, exist := instance.Annotations[common.EnableAgentServiceKey]; exist && enableAgentServiceValue == common.EnableAgentServiceTrue {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:1069]: `EnableAgentServiceKey` -> `EnableServeServiceKey`

**Review by kevin85421 — COMMENTED** [review]
Thank you for the contributions! I've left some comments. Would you or someone from IBM/Red Hat mind manually testing this PR to ensure that the Kubernetes service is created successfully and selects the correct Ray Pods after addressing the comments?

File: ray-operator/controllers/ray/common/pod.go (hunk 1 above)
Comment (blublinsky) [reply ray-operator/controllers/ray/common/pod.go:261]: fixed

File: ray-operator/controllers/ray/common/service.go (hunk 2 above)
Comment (blublinsky) [reply ray-operator/controllers/ray/common/service.go:154]: fixed

File: ray-operator/controllers/ray/common/service.go (hunk 3 above)
Comment (blublinsky) [reply ray-operator/controllers/ray/common/service.go:174]: It is used as selector. No harm in using it

File: ray-operator/controllers/ray/common/service.go (hunk 4 above)
Comment (blublinsky) [reply ray-operator/controllers/ray/common/service.go:165]: fixed

File: ray-operator/controllers/ray/common/service_test.go (hunk 5 above)
Comment (blublinsky) [reply ray-operator/controllers/ray/common/service_test.go:461]: I do not see any harm in it

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 6 above)
Comment (blublinsky) [reply ray-operator/controllers/ray/raycluster_controller.go:341]: Sorry,  `EnableAgentServiceKey` is the only one defined

File: ray-operator/controllers/ray/raycluster_controller.go (hunk 7 above)
Comment (blublinsky) [reply ray-operator/controllers/ray/raycluster_controller.go:1069]: Same as above

**Comment by blublinsky** [comment]
The error here is due to the age of the PR. I can rebase it, but this will bloat the PR with a lot of things, that are not relevant

**Comment by kevin85421** [comment]
The commit https://github.com/ray-project/kuberay/pull/1582/commits/8bfc8d6431d7975656893cce6c27ca5e276a1dae only has a new doc. Maybe you forgot to `git add` some changes to this repo?

> I can rebase it, but this will bloat the PR with a lot of things, that are not relevant

If you rebase it correctly, it should not bloat with irrelevant things.

**Comment by blublinsky** [comment]
@kevin85421 PR is rebased.

**Comment by kevin85421** [comment]
Your rebase process seems to have some issues. It adds a lot of irrelevant things to this PR. Maybe we can open a new PR instead?

**Comment by blublinsky** [comment]
Thats how rebase works, sorry. These additional things are completely harmless, as they come from git. They are already there

**Comment by kevin85421** [comment]
I guess you did not resolve the conflicts correctly before continuing with `rebase --continue`, but it does not matter. Would you mind opening a new PR instead? I cannot merge this PR with a lot of irrelevant things. If you do not want to do that, I can open a PR to implement this when I have time. Thanks!

**Pull request closed without merging**